errdefs/errdefs.go

```diff
@@ -0,0 +1,31 @@
+package errdefs
+
+import "errors"
+
+type internalErr struct {
+	error
+}
+
+func (internalErr) System() {}
+
+func (err internalErr) Unwrap() error {
+	return err.error
+}
+
+type system interface {
+	System()
+}
+
+var _ system = internalErr{}
+
+func Internal(err error) error {
+	if err == nil {
+		return nil
+	}
+	return internalErr{err}
+}
+
+func IsInternal(err error) bool {
+	var s system
+	return errors.As(err, &s)
```

Comment:
Could this just be `errors.Is`? Also, does system need to implement err for this to work?

Reply:
No, `Is` would do a direct comparison of any error in chain.

If you mean adding `func (e internalErr) Is(target) bool { ... interface check }` and then calling `errors.Is(err, InternalErr{})` instead, then this will only work if some error in the chain is type `internalErr` and not when there is any error that just satisfies system interface as otherwise the custom `Is()` method would not get called.

> Also, does system need to implement err for this to work?

No. Err needs to be assignable to system. System does not need to be assignable to error.

Reply:
Ok, I see.

Typically, I would expect the following to be true for an internal error:

```
err := fallableWithInternalError()
errors.Is(err, errdefs.ErrInternal)
```

We'd likely need to have `Unwrap() []error` to make this work with a causal chain but that is why I was asking about having the pointer in the internalErr type above.

I guess it would have to be something like this:

```go
type causal struct {
	err   error
	cause error
}

func (c causal) Error() string {
	return fmt.Sprintf("%v: %v", c.err, c.cause)
}

func (c causal) Unwrap() []error {
	errs := [2]error{c.err, c.cause}
	return errs[:]
}

type errInternal struct{}

var ErrInternal = errInternal{}

func (errInternal) Error() string { return "internal error" }

func WithCause(err, cause error) error {
	return causal{
		err:   err,
		cause: cause,
	}
}

func (errInternal) System() {}

type system interface {
	System()
	Error() string
}

func main() {
	var s system
	err := WithCause(ErrInternal, fmt.Errorf("the disk is done or something"))
	fmt.Println(errors.As(err, &s))
	fmt.Println(errors.Is(err, ErrInternal))
}
```

I don't think we need to block the PR on this.

Reply:
The main difference is still that for `errors.Is(err, ErrInternal)` this would only work if `errdefs.Internal()` (or some form of `Wrap(errdefs.ErrInternal)` is called on the error and not if the system interface is implemented or not. So would not automatically pick up containerd and moby system errors.